### PR TITLE
fix for `spack stage` command not extracting packages in custom paths

### PR DESCRIPTION
**Pull request by albestro**
Using `spack stage -p`, i.e. specifying a custom path, the stage directory gets created but the package is not extracted.

Apparently, the problem is that the stage directory was getting created "manually" by the command, which alters the logic of the `do_stage()` command and the extraction itself.

I went for the more straightforward solution as far as I can understand, but I'm looking forward to feedbacks from more experienced and knowledged spack community members. Thanks in advance!

### CODE REVIEW

**Comment by albestro** [comment]
Thanks to the failing test I may have realized another bit of the story.

### TL;DR

`spack stage -p` forces a creation of `<custom-staging-dir>/spack-src` folder, which is also used as condition to determine if the package has been already extracted or not. Since it is created before calling `do_stage`, when a custom path is specified, the package does not get extracted.

The test is currently failing because `do_stage` is replaced with the mock `fake_stage` which expects the path to be created, but with the current change in this PR the staging dir would be created by `do_stage`.

This PR is strictly related to #23963. Since I don't know exactly the problem it was addressing, I'm going to ask feedbacks from who was involved there.

### Staging

Staging of a package is done via `Package.do_stage()` which triggers in order:
1. stage creation
2. package fetching
3. package extraction

as can be seen from the code here
https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/package.py#L1547-L1562

### Stage creation

This is `stage.create()`, which is responsible of creating the staging dir (which we may want to specify via the command line with `spack stage -p`).
https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/stage.py#L620-L633

On branch `develop`, `spack stage` command creates the staging dir beforehand using the free function `create_stage_root`.

https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/cmd/stage.py#L33-L34

I don't know exactly the rationale behind this function, but it is a bit different from `Stage.create()`. Indeed, it has this additional behavior (added in #23963, keep in mind it since it is relevant for this PR):

https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/stage.py#L100-L106

Creating `spack-src` folder inside staging dir has a "side-effect" on the logic of staging, in particular about package extraction. Keep reading.

### Package extraction

Let's see what point 3 does with `self.stage.expand_archive()`:
https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/stage.py#L604-L612

As it can be seen, package expansion is subordinated to the condition `self.expanded`

https://github.com/spack/spack/blob/6ecee4e6d51a1c750a526c11c11dfd24162afc9e/lib/spack/spack/stage.py#L402-L405

which checks if `spack-src` already exists for assuming that the package has been already extracted.

**Comment by vsoch** [comment]
@albestro I want to suggest you get your tests working before you request reviews! You will undoubtably need to fix/change code to get them working and then ask for review again, so it makes sense to get them working first. Thank you!

**Comment by albestro** [comment]
> @albestro I want to suggest you get your tests working before you request reviews! You will undoubtably need to fix/change code to get them working and then ask for review again, so it makes sense to get them working first. Thank you!

Thanks for the feedback @vsoch. I didn't fix it immediately because it would mean changing the test.

Probably I should have just tagged all of you without asking the reviews, but the point is that I'd like to get a feedback from you before changing the test.

I would remove this line:
https://github.com/spack/spack/blob/653ed786455b2ed25576a5bf14178059bbfe1b05/lib/spack/spack/test/cmd/stage.py#L44

**Comment by vsoch** [comment]
I'm not the right person to review this, not super familiar with how this works sorry @tldahlgren

**Comment by haampie** [comment]
@cosmicexplorer can you have a look?

**Review by haampie — APPROVED** [review]
